drivers: intc: plic: add support for multiple instances

ycsin opened the pull request:
1. Refactored the build_all test added in #62541
2. Refactored the `intc_plic` driver to be multi-instance ready
3. Updated the `sw_isr_*` so that it also records the device pointer of the interrupt controllers & provide functions to do a lookup for the controller using the IRQN, as well as getting the ISR table offset for a controller
4. Updated the `intc_plic` driver to use the functions in the `sw_isr` so that it can differentiate parent controller from IRQN, and invoke the correct ISR with the correct offset from `sw_isr_*`
   - Modified some description in the `riscv_plic` header file, previously the driver expects the trimmed IRQN returned by `irq_from_level_2` as there can only be one controller, with multi-instance, we need the full multi-level-encoded IRQN, so that the driver can determine the right controller with `z_get_sw_isr_device` using the 1st level IRQ.
6. Added a build_all test for multi-instance on `qemu_riscv*` as no QEMU machine supports more than one PLIC. Not sure if this is helpful.

fixes #62441
fixes #62956
fixes #63465

ycsin committed:
drivers: plic: support multiple instances for multi-level

Most of the public APIs in `riscv_plic.h`
(except `riscv_plic_get_irq` & `riscv_plic_get_dev`) expect the
`irq` argument to be in Zephyr-encoded format, instead of the
previously `irq_from_level_2`-stripped version. The first level
IRQ is needed by `intc_plic` to differentiate between the
parent interrupt controllers, so that correct ISR offset can be
obtained using the LUT in `sw_isr_common`.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin committed:
drivers: intc: plic: initial refactor for multi-instance support

Use a config struct to store per-instance device config during
init and connect the IRQ based on the devicetree instead of
hardcoded value and instance number.

The `get_plic_dev_from_irq` is still a placeholder for now and
always return the first instance.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin committed:
tests: build_all: plic: multi instance

Add PLIC multi-instance build-only test.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

cfriedt commented on drivers/interrupt_controller/intc_plic.c:
```diff
@@ -146,54 +187,58 @@ void riscv_plic_set_priority(uint32_t irq, uint32_t priority)
  * @return PLIC-specific interrupt line causing an interrupt.
  */
-int riscv_plic_get_irq(void)
+unsigned int riscv_plic_get_irq(const struct device *dev)
```
It would be better to add a new function called something like `uint32_t riscv_plic_get_irq_by_dev(const struct device *dev)` instead of modifying the signature of the existing code.

With that, `unsigned int riscv_plic_get_irq(void)` can become a 1-line wrapper around the new function (passing `NULL`) and fewer 1-line changes are needed in e.g. unrelated gpio drivers

ycsin replied:
fixed

cfriedt commented on dts/bindings/interrupt-controller/sifive,plic-1.0.0.yaml:
```diff
@@ -12,7 +12,7 @@ properties:
-    default: 4224
-    description: Offset of the trigger type register if supported
+
+  support-edge-interrupt:
```
Is just a feature that is always available? Do we even need a Devicetree property here? If it's always available, then the Devicetree property is superfluous (and should probably just be removed).

cfriedt replied:
@fkokosinski - can you weigh in here?

fkokosinski replied:
Hi @cfriedt. I've looked around a bit and found out that in the Linux irqchip driver implementation for PLIC SiFive platforms (fu540, fu740; HiFive Unleashed and HiFive Unmatched, respectively) also don't support the edge mode: https://github.com/torvalds/linux/blob/master/drivers/irqchip/irq-sifive-plic.c#L565-L581.

This IMO really suggests that it's not a bug in the documentation that SiFive provides for their chips, and that they really lack the support of edge-triggered interrupts, so I would suggest to  leave the `support-edge-interrupt` property as currently defined by this PR.

cfriedt replied:
@keith-packard  - any insider info here from your previous role?

cfriedt replied:
@fkokosinski

https://www.kernel.org/doc/Documentation/devicetree/bindings/interrupt-controller/sifive%2Cplic-1.0.0.yaml

Seems to suggest that SiFive's PLIC *does* support edge triggering in the kernel (and that all of them do), and it was signed-off by Paul & Palmer, so I'm inclined to lean toward going with it.

> The PLIC supports both edge-triggered and level-triggered interrupts

What it does say is different in some implementations is whether interrupts are queued or dropped if triggered while they are already pended, which can probably be handled in code, in a somewhat generic way. Good to be aware of, in any case.

keith-packard replied:
> @keith-packard - any insider info here from your previous role?

Afraid I'd have to read through documentation I no longer have access to -- I don't specifically remember anything about PLIC interrupt configurations...

cfriedt replied:
> Hi @cfriedt. I've looked around a bit and found out that in the Linux irqchip driver implementation for PLIC SiFive platforms (fu540, fu740; HiFive Unleashed and HiFive Unmatched, respectively) also don't support the edge mode: https://github.com/torvalds/linux/blob/master/drivers/irqchip/irq-sifive-plic.c#L565-L581.

Yea, that's the thing - some datasheets do, and some datasheets don't, but the memory map in Linux does support the register, and it's a "quirk" in Linux. Andes and telink support the register and document it in their datasheet. SiFive supports it in their SoC's, but it's undocumented / "reserved", almost as if it were a kind of copy-paste issue from the PLIC spec. 

> SiFive SOCs include an implementation of the Platform-Level Interrupt Controller (PLIC)..  the PLIC supports both edge-triggered and level-triggered interrupts..

http://archive.lwn.net:8080/linux-kernel/b4ddcb20-6109-22ac-4691-c2c012ea112a@wdc.com/T/

I very much suspect that it has been a bit of a contentious topic. The spec was (unintentionally) ambiguous and it's possible that some vendors had one interpretation, while others had another interpretation.

> This IMO really suggests that it's not a bug in the documentation that SiFive provides for their chips, and that they really lack the support of edge-triggered interrupts, so I would suggest to leave the `support-edge-interrupt` property as currently defined by this PR.

https://github.com/torvalds/linux/blob/master/drivers/irqchip/irq-sifive-plic.c#L565-L581.

Of course, we now know that the opposite is true, as we have read. SiFive does, in fact, support edge triggering (via a "quirk"), even though their datasheet does not mention it in the memory map, which is the same "documentation bug" in the PLIC spec itself. Even the ESP32-C3 interrupt controller supports both edge and level-triggered interrupts.

Also important to note, is that `riscv,plic0` and `sifive,plic-1.0.0` are deprecated bindings for "legacy" systems, while all newer compatibles are hard-coded to support edge triggering and use a binding that also hard-codes this support. I would say, that this was a chosen "documentation bug" workaround rather than revising the already-ratified PLIC spec.

https://github.com/torvalds/linux/blob/master/drivers/irqchip/irq-sifive-plic.c#L565-L581

So, IMHO, I think this was a bit of a "documentation bug" that unfortunately caused some misinterpretations of the specification and likely had some undesirable fallout. It happens.

ycsin replied:
The parameter is dropped in #63273

cfriedt commented on tests/drivers/build_all/interrupt_controller/intc_plic/app.edge_interrupt.overlay:
```diff
@@ -7,7 +7,7 @@
 	soc {
 		plic: interrupt-controller@c000000 {
-			riscv,trigger-reg-offset = < 0x00001080 >;
+			support-edge-interrupt;
```
possibly remove, depending on what the consensus is.

ycsin replied:
The parameter is dropped in #63273

cfriedt commented on drivers/interrupt_controller/intc_plic.c:
```diff
@@ -46,17 +76,20 @@ static int save_irq;
-		trig += (irq >> PLIC_EDGE_TRIG_SHIFT);
-		return *trig & BIT(irq);
+		trig += (local_irq >> PLIC_EDGE_TRIG_SHIFT);
+		return *trig & BIT(local_irq);
```
should use `sys_write32(..)`

ycsin replied:
#62956

cfriedt commented on drivers/interrupt_controller/intc_plic.c:
```diff
@@ -73,8 +106,10 @@ static int riscv_plic_is_edge_irq(uint32_t irq)
 {
+	const struct device *dev = get_plic_dev_from_irq(irq);
+	const struct plic_config *config = dev->config;
+	volatile uint32_t *en = (volatile uint32_t *) config->irq_en;
```
will also need to use `sys_read32(..)` / `sys_write32(..)`

Please file a bug for that too.

ycsin replied:
#62956

cfriedt commented on drivers/interrupt_controller/intc_plic.c:
```diff
@@ -202,49 +247,72 @@ static void plic_irq_handler(const void *arg)
-	    (volatile struct plic_regs_t *)PLIC_REG;
-	int i;
+	const struct plic_config *config = dev->config;
+	volatile uint32_t *en = (volatile uint32_t *) config->irq_en;
```
Also `sys_read32()` / `sys_write32()`

ycsin replied:
#62956

ycsin commented on dts/bindings/interrupt-controller/interrupt-controller.yaml:
```diff
@@ -12,3 +12,6 @@ properties:
     type: int
     required: true
     description: Number of items to expect in an interrupt specifier
+  "has-multi-instance":
```
This is needed for the DT macro in `sw_isr_common.c` to get the device pointer of a root interrupt controller.
Alternatively this might be replaceable by Kconfigs (`INTC_HAS_MULTI_INSTANCE`, selected by interrupt controller that supports multi-instance, but not necessarily enabled) to only compile the multi-instance related code when `INTC_ENABLE_MULTI_INSTANCE` is enabled

cfriedt replied:
Consistent with Zephyr's driver model, it's better to leave the gate assuming that drivers can be multi-instance. That's also a DT thing, and not a Kconfig thing. The number of instances of a dt-compat is implicitly part of DT. So I would say that it's better to proceed under the assumption that this is always `true`. There is no real need to reiterate something explicitly that is implicit in devicetree.

If this is part of the `sw_isr_common.c`, please make the necessary adjustments to the other PR.

ycsin replied:
this prop is removed in #63172

ycsin committed:
drivers: intc: plic: make sense of magic number

Added some defines and helper functions to help with the
arithmetics so that the bit shifts and stuff do not look like
magic number.

Converted manual bit shift/set/unset to use macros provided by
Zephyr.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin committed:
drivers: intc: plic: Use sys IO APIs to access memory-mapped registers

Use arch-specific sys IO APIs to access the memory-mapped
registers to ensure safe memory operations

fixes #62956

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin commented:
rebased

carlocaione commented on drivers/interrupt_controller/intc_plic.c:
```diff
@@ -22,12 +22,15 @@
 #include <zephyr/irq.h>
 
 #define PLIC_BASE_ADDR(n) DT_INST_REG_ADDR(n)
+#define PLIC_REG_PRIO_OFFSET 0x0
```
should this be `0x1000`?

ycsin replied:
https://github.com/zephyrproject-rtos/zephyr/blob/a2b238eee77cf9137db858618eb2881c53963330/dts/riscv/virt.dtsi#L168-L171

prio offset should be 0, 0x1000 is the offset of pending bits according to the specs

carlocaione commented on drivers/interrupt_controller/intc_plic.c:
```diff
@@ -22,12 +22,15 @@
+#define PLIC_REG_PRIO_OFFSET 0x0
 #define PLIC_REG_TRIG_TYPE_OFFSET 0x1080
+#define PLIC_REG_IRQ_EN_OFFSET 0x2000
+#define PLIC_REG_REGS_OFFSET 0x200000
```
Can you use one single alignment style?

ycsin replied:
These macros will have a single alignment style in the end, is that acceptable? I didnt adopt the existing alignment style as it tends to go out of alignment in the long run

carlocaione commented on arch/common/sw_isr_common.c:
I think the changes in this file should be submitted as a standalone PR because it is not strictly related to PLIC at all.

cfriedt replied:
That's probably a healthy thing. If it can be done independently, lets do that

cfriedt commented:
Going to add `DNM` here until #63172 is merged. Feel free to review in any case.

cfriedt reviewed (dismissed):
Pre-emptively acking. Please remember to rebase once #63172 is merged (for sw_isr_common changes)

ycsin commented on arch/common/sw_isr_common.c:
```diff
@@ -16,12 +16,38 @@
+
+/*
+ * Expands to node_id if its IRQN is equal to `irq`, nothing otherwise
+ * This only works for `irq` between 0 & 255, see `IS_EQ`
```
This will support up to 4095 after #63496 is merged

jhedberg commented:
> Pre-emptively acking. Please remember to rebase once #63172 is merged (for sw_isr_common changes)

#63172 wasn't flagged for 3.5 even though this PR is. I fixed the milestone for the other PR, but it's getting quite late in the game for 3.5 to introduce such big changes.

ycsin commented:
rebased on top of #63172, following the changes over there
